benchmark/common_utils.hpp

```diff
@@ -5,15 +5,147 @@
 #include <benchmark/benchmark.h>
 #include <chrono>
 #include <climits>
+#include <fstream>
+#include <functional>
+#include <iostream>
 #include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
 
+#include "benchmark_cli_args.hpp"
 #include "blas_meta.h"
 
 using index_t = int;
+using blas1_param_t = int;
+using blas2_param_t = std::tuple<std::string, int, int>;
+using blas3_param_t = std::tuple<std::string, std::string, int, int, int>;
+
+namespace blas_benchmark {
 
-namespace benchmark {
 namespace utils {
 
+/**
+ * @fn parse_csv_file
+ * @brief Returns a vector containing the parameters for a benchmark as tuples,
+ * read from the given csv file
+ */
+template <typename param_t>
+std::vector<param_t> parse_csv_file(
+    std::string& filepath,
+    std::function<param_t(std::vector<std::string>&)> func) {
+  std::vector<param_t> csv_data;
+  std::ifstream data(filepath);
+  std::string line;
+  while (std::getline(data, line)) {
+    if (line.empty()) continue;
+    line.push_back({});
+    std::stringstream lineStream(line);
+    std::string cell;
+    std::vector<std::string> csv_line;
+    while (std::getline(lineStream, cell, ',')) {
+      csv_line.push_back(cell);
+    }
+    csv_data.push_back(func(csv_line));
+  }
+  return csv_data;
+}
+
+/**
+ * @fn get_range
+ * @brief Prints a warning to the user if no CSV parameter file has been given
+ */
+inline void warning_no_csv() {
+  std::cerr
+      << "WARNING: no CSV parameter file has been given. Default ranges will "
+         "be used."
+      << std::endl;
+}
+
+/**
+ * @fn get_range
+ * @brief Returns a range containing the parameters, either read from a file
+ * according to the command-line args, or the default ones.
+ * This function must be implemented for each blas level.
+ */
+template <typename param_t>
+std::vector<param_t> get_params(Args& args);
+
+template <>
+inline std::vector<blas1_param_t> get_params<blas1_param_t>(Args& args) {
+  if (args.csv_param.empty()) {
+    warning_no_csv();
+    std::vector<blas1_param_t> blas1_default;
+    for (int size = 4096; size <= 1048576; size *= 2)
+      blas1_default.push_back(size);
+    return blas1_default;
+  } else
+    return parse_csv_file<blas1_param_t>(
+        args.csv_param,
+        [&](std::vector<std::string>& v) { return std::stoi(v[0]); });
+}
+
+template <>
+inline std::vector<blas2_param_t> get_params<blas2_param_t>(Args& args) {
+  if (args.csv_param.empty()) {
+    warning_no_csv();
+    std::vector<blas2_param_t> blas2_default;
+    constexpr int dmin = 64, dmax = 1024;
+    for (std::string t : {"n", "t"})
+      for (int m = dmin; m <= dmax; m *= 2)
+        for (int n = dmin; n <= dmax; n *= 2)
+          blas2_default.push_back(std::make_tuple(t, m, n));
+    return blas2_default;
+  } else
+    return parse_csv_file<blas2_param_t>(
+        args.csv_param, [&](std::vector<std::string>& v) {
+          return std::make_tuple(v[0].c_str(), std::stoi(v[1]),
```

Comment:
How will this work if only 0 or 1 values were parsed from the csv?

Reply:
I think we discussed about checking the CSV format or not. It adds a lot of code to check it properly and provide relevant information about the failure, but what we can do, maybe, is to catch the exceptions, print a message, and then either fail or use defaults. Any opinion on this?

Reply:
I'm not really bothered about having a perfectly crafted error message. Perhaps parse_csv_file or the param_t could have the number of fields per record, and it just says "invalid csv file!" if they don't match. I'd rather not have an invalid csv file causing UB.

Reply:
I have added some basic checking with 3 possible messages (+ printing the line where it fails):
* `invalid number of parameters (x expected)`
* `invalid parameter`
* `unrecognized transpose type: x`